rclpy/src/rclpy/publisher.cpp

```diff
@@ -75,60 +48,52 @@ publisher_create(
   }
 
-  // Use smart pointer to make sure memory is free'd on error
-  auto deleter = [](rclpy_publisher_t * ptr) {_rclpy_destroy_publisher(ptr);};
-  auto pub = std::unique_ptr<rclpy_publisher_t, decltype(deleter)>(
-    static_cast<rclpy_publisher_t *>(PyMem_Malloc(sizeof(rclpy_publisher_t))),
-    deleter);
-  if (!pub) {
-    throw std::bad_alloc();
-  }
-  pub->publisher = rcl_get_zero_initialized_publisher();
-  pub->node = node;
+  rcl_publisher_ = std::shared_ptr<rcl_publisher_t>(
+    new rcl_publisher_t,
+    [this](rcl_publisher_t * publisher)
+    {
+      auto node = node_handle_->cast_or_warn<rcl_node_t *>("rcl_node_t");
+
+      rcl_ret_t ret = rcl_publisher_fini(publisher, node);
+      if (RCL_RET_OK != ret) {
+        // Warning should use line number of the current stack frame
+        int stack_level = 1;
+        PyErr_WarnFormat(
+          PyExc_RuntimeWarning, stack_level, "Failed to fini publisher: %s",
+          rcl_get_error_string().str);
+        rcl_reset_error();
+      }
+      delete publisher;
+    });
+
+  *rcl_publisher_ = rcl_get_zero_initialized_publisher();
 
   rcl_ret_t ret = rcl_publisher_init(
-    &(pub->publisher), node, msg_type,
+    rcl_publisher_.get(), node, msg_type,
     topic.c_str(), &publisher_ops);
-  if (ret != RCL_RET_OK) {
-    if (ret == RCL_RET_TOPIC_NAME_INVALID) {
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_TOPIC_NAME_INVALID == ret) {
       std::string error_text{"Failed to create publisher due to invalid topic name '"};
       error_text += topic;
       error_text += "'";
       throw py::value_error(error_text);
     }
     throw RCLError("Failed to create publisher");
   }
+}
 
-  PyObject * pypub_c =
-    rclpy_create_handle_capsule(pub.get(), "rclpy_publisher_t", _rclpy_destroy_publisher);
-  if (!pypub_c) {
-    throw py::error_already_set();
-  }
-  auto pypub = py::reinterpret_steal<py::capsule>(pypub_c);
-  // pypub now owns the rclpy_publisher_t
-  pub.release();
-
-  auto pub_handle = static_cast<rclpy_handle_t *>(pypub);
-  auto node_handle = static_cast<rclpy_handle_t *>(pynode);
-  _rclpy_handle_add_dependency(pub_handle, node_handle);
-  if (PyErr_Occurred()) {
-    _rclpy_handle_dec_ref(pub_handle);
-    throw py::error_already_set();
-  }
-
-  return pypub;
+void Publisher::destroy()
+{
+  rcl_publisher_.reset();
+  node_handle_.reset();
 }
 
 const char *
-publisher_get_logger_name(py::capsule pypublisher)
+Publisher::get_logger_name()
 {
-  auto pub = static_cast<rclpy_publisher_t *>(
-    rclpy_handle_get_pointer_from_capsule(pypublisher.ptr(), "rclpy_publisher_t"));
-  if (!pub) {
-    throw py::error_already_set();
-  }
+  auto node = node_handle_->cast<rcl_node_t *>("rcl_node_t");
 
-  const char * node_logger_name = rcl_node_get_logger_name(pub->node);
+  const char * node_logger_name = rcl_node_get_logger_name(node);
   if (!node_logger_name) {
     throw RCLError("Node logger name not set");
   }
```

Comment:
@ahcorde hmm why does `Publisher::get_logger_name()` throw on error, whereas `Subscription::get_logger_name()` returns `None`? I'd rather be consistent.